File: 백준/Gold/1525. 퍼즐/퍼즐.java
import java.util.*;
import java.io.*;

public class Main {

    public static void main(String[] args) throws IOException{
    	BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    	board = new int[3][3];
    	sy = sx = -1;
    	int tmp[][] = new int[3][3];
    	for(int y=0; y<3; y++) {
    		st = new StringTokenizer(br.readLine());
    		for(int x =0 ; x<3; x++) {
    			board[y][x] = Integer.parseInt(st.nextToken());
    			tmp[y][x] = board[y][x];
    			if(board[y][x] == 0) {
    				sy = y;
    				sx = x;
    			}
    				
    		}
    			
    	}
    	
    	State start = new State(sy,sx, board);
    	Queue<State> q = new ArrayDeque<>();
    	Map<State, Integer> visited = new HashMap<>();
    	q.add(start);
    	visited.put(start, 0);
    	
    	while(!q.isEmpty()) {
    		State cur = q.poll();
    		
    		for(int dir = 0; dir < 4; dir++) {
    			int ny = cur.y + dy[dir];
    			int nx = cur.x + dx[dir];
    			
    			if(OOB(ny,nx))
    				continue;
    			
    			State nxt = swap(cur.y, cur.x, ny,nx, cur.board);
    			if(visited.containsKey(nxt))
    				continue;
    			
    			visited.put(nxt, visited.get(cur) + 1);
    			q.add(nxt);
    		}
    	}
    	
    	int temp[][] = new int[3][3];
    	for(int y=0;y<3; y++) {
    		for(int x=0; x<3; x++) {
    			temp[y][x] = y * 3 + x + 1;
    		}
    	}
    	temp[2][2] = 0;
    	State end = new State(2,2,temp);
    	System.out.println(visited.getOrDefault(end, -1));
    }
    
    static State swap(int curY, int curX, int ny, int nx, int board[][]) {
    	int ret[][] = new int[3][3];
    	for(int y=0;y <3; y++) {
    		for(int x = 0; x< 3; x++) {
    			ret[y][x] = board[y][x];
    		}
    	}
    	
    	int temp = board[ny][nx];
    	ret[ny][nx] = 0;
    	ret[curY][curX] = temp;
    	
    	return new State(ny,nx, ret);
    }
    
    static int dy[] = {-1,1,0,0};
    static int dx[] = {0,0,-1,1};
    static boolean OOB(int y,int x) {
    	return y<0 || y>=3 || x<0 || x>=3;
    }
    
    static int board[][];
    static int sy, sx;
    static StringTokenizer st;
    static class State{
    	int y,x;
    	int board[][];
    	
    	public State(int y, int x, int board[][]) {
    		this.y = y;
    		this.x = x;
    		this.board = board;
    	}
    	
    	public boolean equals(Object o) {
    		if(this == o)
    			return true;
    		
    		if(o == null || getClass()!=o.getClass())
    			return false;
    		
    		State tmp = (State)o;
    		return Arrays.deepEquals(board, tmp.board);
    		
    	}
    	
    	public int hashCode() {
    		return Arrays.deepHashCode(board);
    	}
    	
    }
    
}

File: 백준/Gold/1525. 퍼즐/퍼즐.py
from collections import deque

def solve():
    board = ""
    sy, sx = -1, -1
    for y in range(3):
        row = input().split()
        for x in range(3):
            if row[x] == '0':
                sy, sx = y, x
            board += row[x]

    start = (sy, sx, board)
    q = deque([start])
    visited = {start[2]}  # 문자열을 저장하는 set
    dist = {start[2]: 0}

    dy = [-1, 1, 0, 0]
    dx = [0, 0, -1, 1]

    def oob(y, x):
        return y < 0 or y >= 3 or x < 0 or x >= 3

    def swap(cur_y, cur_x, ny, nx, board_str):
        board_list = list(board_str)
        cur_idx = cur_y * 3 + cur_x
        nxt_idx = ny * 3 + nx
        board_list[cur_idx], board_list[nxt_idx] = board_list[nxt_idx], board_list[cur_idx]
        return (ny, nx, "".join(board_list))

    while q:
        cur_y, cur_x, cur_board = q.popleft()

        if cur_board == "123456780":
            print(dist[cur_board])
            return

        for i in range(4):
            ny, nx = cur_y + dy[i], cur_x + dx[i]

            if oob(ny, nx):
                continue

            nxt = swap(cur_y, cur_x, ny, nx, cur_board)
            if nxt[2] in visited:
                continue

            visited.add(nxt[2])
            dist[nxt[2]] = dist[cur_board] + 1
            q.append(nxt)

    print(-1)

if __name__ == "__main__":
    solve()